File: src/Servers/ServerBrowser/src/V2/Aggregate/ServerInfoBuilder.cs
using System;
using System.Collections.Generic;
using System.Net;
using UniSpy.Server.QueryReport.V2.Aggregate.Redis.GameServer;
using UniSpy.Server.ServerBrowser.V2.Application;
using UniSpy.Server.ServerBrowser.V2.Enumerate;

namespace UniSpy.Server.ServerBrowser.V2.Aggregate
{
    public static class ServerInfoBuilder
    {

        /// <summary>
        /// Add more server info here
        /// the sequence of server info is important
        /// 1.PRIVATE_IP_FLAG length=4
        /// 2.ICMP_IP_FLAG length=4
        /// 3.NONSTANDARD_PORT_FLAG length=2
        /// 4.NONSTANDARD_PRIVATE_PORT_FLAG length=2
        /// </summary>
        /// <param name="flag"></param>
        /// <param name="serverInfo"></param>
        public static List<byte> BuildServerInfoHeader(GameServerFlags flag, GameServerInfo serverInfo)
        {
            List<byte> header = new List<byte>();
            // add key flag
            header.Add((byte)flag);
            // we add server public ip here
            header.AddRange(serverInfo.HostIPAddress.GetAddressBytes());
            // we check host port is standard port or not
            CheckNonStandardPort(header, serverInfo);
            // check if game can directly query information from server
            CheckUnsolicitedUdp(header, serverInfo);
            // we check the natneg flag
            CheckNatNegFlag(header, serverInfo);
            // now we check if there are private ip
            CheckPrivateIP(header, serverInfo);
            // we check private port here
            CheckNonStandardPrivatePort(header, serverInfo);
            // we check icmp support here
            CheckICMPSupport(header, serverInfo);

            // _serverListData.AddRange(header);
            return header;
        }
        public static void CheckNatNegFlag(List<byte> header, GameServerInfo serverInfo)
        {
            if (serverInfo.ServerData.ContainsKey("natneg"))
            {
                var natNegFlag = int.Parse(serverInfo.ServerData["natneg"]);
                var unsolicitedUdp = header[0] & (byte)GameServerFlags.UnsolicitedUdpFlag;
                if (natNegFlag == 1 && unsolicitedUdp == 0)
                {
                    header[0] ^= (byte)GameServerFlags.ConnectNegotiateFlag;
                }
            }
        }
        public static void CheckUnsolicitedUdp(List<byte> header, GameServerInfo serverInfo)
        {
            if (serverInfo.ServerData.ContainsKey("allow_unsolicited_udp"))
            {
                var unsolicitedUdp = int.Parse(serverInfo.ServerData["unsolicitedudp"]);
                if (unsolicitedUdp == 1)
                {
                    header[0] ^= (byte)GameServerFlags.UnsolicitedUdpFlag;
                }
            }
        }
        /// <summary>
        /// !when game create a channel chat, it will use both the public ip and private ip to build the name.
        /// !Known game: Worm3d
        /// </summary>
        public static void CheckPrivateIP(List<byte> header, GameServerInfo serverInfo)
        {
            if (QueryReport.Application.StorageOperation.PeerGroupList.ContainsKey(serverInfo.GameName))
            {
                // We already have the localip. Bytes are worng.
                if (serverInfo.ServerData.ContainsKey("localip0"))
                {
                    header[0] ^= (byte)GameServerFlags.PrivateIPFlag;
                    // there are multiple localip in dictionary we do not know which one is needed here,
                    // so we just send the first one.
                    byte[] bytesAddress = IPAddress.Parse(serverInfo.ServerData["localip0"]).GetAddressBytes();
                    header.AddRange(bytesAddress);
                }
            }
        }
        public static void CheckNonStandardPort(List<byte> header, GameServerInfo serverInfo)
        {
            // !! only dedicated server have different query report port and host port
            // !! but peer server have same query report port and host port
            // todo we have to check when we need send host port or query report port
            if (serverInfo.QueryReportPort != ClientInfo.QueryReportDefaultPort)
            {
                header[0] ^= (byte)GameServerFlags.NonStandardPort;
                byte[] htonPort = serverInfo.QueryReportPortBytes;
                header.AddRange(htonPort);
            }
        }
        public static void CheckNonStandardPrivatePort(List<byte> header, GameServerInfo serverInfo)
        {
            // we check private port here
            if (serverInfo.ServerData.ContainsKey("localport"))
            {
                if (serverInfo.ServerData["localport"] != ""
                && serverInfo.ServerData["localport"] != ClientInfo.QueryReportDefaultPort.ToString())
                {
                    header[0] ^= (byte)GameServerFlags.NonStandardPrivatePortFlag;
                    byte[] port = BitConverter.GetBytes(short.Parse(serverInfo.ServerData["localport"]));
                    header.AddRange(port);
                }
            }
        }
        public static void CheckICMPSupport(List<byte> header, GameServerInfo serverInfo)
        {
            if (serverInfo.ServerData.ContainsKey("icmp_address"))
            {
                header[0] ^= (byte)GameServerFlags.IcmpIpFlag;
                byte[] bytesAddress = IPAddress.Parse(serverInfo.ServerData["icmp_address"]).GetAddressBytes();
                header.AddRange(bytesAddress);
            }
        }

    }
}

File: src/frontends/gamespy/protocols/server_browser/v2/aggregations/server_info_builder.py
from socket import inet_aton

from frontends.gamespy.protocols.query_report.aggregates.game_server_info import GameServerInfo

# from backends.protocols.gamespy.query_report.data import get_all_groups
from frontends.gamespy.protocols.server_browser.v2.aggregations.enums import GameServerFlags

# PEER_GROUP_LIST = get_all_groups()
QUERY_REPORT_DEFAULT_PORT = 6500


def build_server_info_header(
    flag: GameServerFlags, server_info: GameServerInfo
) -> bytearray:
    header = bytearray()
    # add key flag
    header.append(flag.value)
    # we add server public ip here
    header.extend(inet_aton(server_info.host_ip_address))
    # we check host port is standard port or not
    check_non_standard_port(header, server_info)
    # check if game can directly query information from server
    check_unsolicited_udp(header, server_info)
    # we check the natneg flag
    check_nat_neg_flag(header, server_info)
    # now we check if there are private ip
    check_private_ip(header, server_info)
    # we check private port here
    check_non_standard_private_port(header, server_info)
    # we check icmp support here
    check_icmp_support(header, server_info)

    return header


def check_nat_neg_flag(header: bytearray, server_info: GameServerInfo):
    if "natneg" in server_info.server_data:
        nat_neg_flag = int(server_info.server_data["natneg"])
        unsolicited_udp = header[0] & GameServerFlags.UNSOLICITED_UDP_FLAG.value
        if nat_neg_flag == 1 and unsolicited_udp == 0:
            header[0] ^= GameServerFlags.CONNECT_NEGOTIATE_FLAG.value


def check_unsolicited_udp(header: bytearray, server_info: GameServerInfo):
    if "allow_unsolicited_udp" in server_info.server_data:
        unsolicited_udp = int(server_info.server_data["unsolicitedudp"])
        if unsolicited_udp == 1:
            header[0] ^= GameServerFlags.UNSOLICITED_UDP_FLAG.value


def check_private_ip(header: bytearray, server_info: GameServerInfo):
    #!when game create a channel chat, it will use both the public ip and private ip to build the name.
    #!known game: Worm3d
    # todo
    # if server_info.game_name in PEER_GROUP_LIST:
        if "localip0" in server_info.server_data:
            header[0] ^= GameServerFlags.PRIVATE_IP_FLAG.value
            bytes_address = inet_aton(server_info.server_data["localip0"])
            header.extend(bytes_address)


def check_non_standard_port(header: bytearray, server_info: GameServerInfo):
    # !! only dedicated server have different query report port and host port
    # !! but peer server have same query report port and host port
    # todo we have to check when we need send host port or query report port
    if server_info.query_report_port != QUERY_REPORT_DEFAULT_PORT:
        header[0] ^= GameServerFlags.NON_STANDARD_PORT_FLAG.value
        hton_port = server_info.query_report_port_bytes
        header.extend(hton_port)


def check_non_standard_private_port(header: bytearray, server_info: GameServerInfo):
    if "localport" in server_info.server_data:
        local_port = server_info.server_data["localport"]
        if local_port != "" and local_port != QUERY_REPORT_DEFAULT_PORT:
            header[0] ^= GameServerFlags.NON_STANDARD_PRIVATE_PORT_FLAG.value
            port = int(local_port).to_bytes(2, byteorder="big")
            header.extend(port)


def check_icmp_support(header: bytearray, server_info: GameServerInfo):
    if "icmp_address" in server_info.server_data:
        header[0] ^= GameServerFlags.ICMP_IP_FLAG.value
        bytes_address = inet_aton(server_info.server_data["icmp_address"])
        header.extend(bytes_address)
